xmtp_content_types/src/attachment.rs

```diff
@@ -1,6 +1,103 @@
+use std::collections::HashMap;
+
+use crate::{utils::get_param_or_default, CodecError, ContentCodec};
+use serde::{Deserialize, Serialize};
+
+use xmtp_proto::xmtp::mls::message_contents::{ContentTypeId, EncodedContent};
+
 pub struct AttachmentCodec {}
 
-//. Legacy content type id at https://github.com/xmtp/xmtp-js/blob/main/content-types/content-type-remote-attachment/src/Attachment.ts
+/// Legacy content type id at https://github.com/xmtp/xmtp-js/blob/main/content-types/content-type-remote-attachment/src/Attachment.ts
 impl AttachmentCodec {
+    const AUTHORITY_ID: &'static str = "xmtp.org";
     pub const TYPE_ID: &'static str = "attachment";
 }
+
+impl ContentCodec<Attachment> for AttachmentCodec {
+    fn content_type() -> ContentTypeId {
+        ContentTypeId {
+            authority_id: Self::AUTHORITY_ID.to_string(),
+            type_id: Self::TYPE_ID.to_string(),
+            version_major: 1,
+            version_minor: 0,
+        }
+    }
+
+    fn encode(data: Attachment) -> Result<EncodedContent, CodecError> {
+        let fallback = Self::fallback(&data);
+        let mut parameters = [("mimeType", data.mime_type)]
+            .into_iter()
+            .map(|(k, v)| (k.to_string(), v))
+            .collect::<HashMap<_, _>>();
+
+        if let Some(filename) = data.filename {
+            parameters.insert("filename".to_string(), filename);
+        }
+
+        Ok(EncodedContent {
+            r#type: Some(Self::content_type()),
+            parameters,
+            fallback: Some(fallback),
+            compression: None,
+            content: data.content,
+        })
+    }
+
+    fn decode(encoded: EncodedContent) -> Result<Attachment, CodecError> {
+        let parameters: &HashMap<String, String> = &encoded.parameters;
+
+        Ok(Attachment {
+            filename: parameters.get("filename").map(|f| f.to_string()),
+            mime_type: get_param_or_default(parameters, "mimeType").to_string(),
+            content: encoded.content,
+        })
+    }
+}
+
+impl AttachmentCodec {
+    fn fallback(content: &Attachment) -> String {
+        if let Some(filename) = &content.filename {
+            format!("[Attachment] {filename}")
+        } else {
+            "[Attachment]".to_string()
+        }
+    }
+}
+
+/// The main content type for attachments
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Attachment {
+    /// The filename of the attachment
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub filename: Option<String>,
+
+    /// The MIME type of the attachment
+    pub mime_type: String,
```

Comment:
Does this need a `#[serde(rename)]` annotation, seeing as the original field uses camel case? And similarly on other codecs, e.g. `content_digest` and `content_length` fields in `RemoteAttachment`

Reply:
It doesn't, only because we get the value from a parameter on the message and not from a JSON payload that needs to be parsed by Serde. See line 28 and 51

Reply:
I see, it's just a little confusing working out what is/isn't serialized because there is a serde annotation on the `filename` field. Not a big deal, but would be tempted to either remove that annotation or just add `#serde[(rename_all` everywhere so that there is a clear standard for the next implementer